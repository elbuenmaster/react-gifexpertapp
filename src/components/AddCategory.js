import React, { useState } from "react";
import Proptypes from 'prop-types'
export const AddCategory = ({ setCategorias }) => {
  
  const [inputValue, setInputValue] = useState('');

  const handleInputValue = (e) => {
    setInputValue(e.target.value);
    console.log('se cambio el valor de la cajita','{',e.target.value,'}');
  };

  const handleSubmit = (e) => {
              e.preventDefault();
              console.log("se dispara Submit");
              if (inputValue.trim().length > 2) {
                // setCategorias: se le agrega, usando una funcion CALLBACK que lea el valor del State y le aÃ±ada el nuevo valor
                setCategorias((cats) => [ inputValue,...cats]);
                setInputValue('');
                console.log("Submit exitoso");
                return true;
              }
              console.log("se cancela Submit");
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input type="text" value={inputValue} onChange={handleInputValue}></input>
    </form>
  );
};

AddCategory.propTypes={
    setCategorias: Proptypes.func.isRequired
}