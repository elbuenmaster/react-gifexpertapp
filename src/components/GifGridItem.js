import React from 'react'
import PropTypes from 'prop-types'
export const GifGridItem = ({id,title,url}) => {
    console.log(id)
    return (
        <div className="card animate__fadeIn">
        
           <img src={url} alt={title}></img> 
           <p>{title}</p>       
       
        </div>
    )
}

GifGridItem.propTypes={
    id:PropTypes.number.isRequired,
    title:PropTypes.string.isRequired,
    url:PropTypes.string.isRequired,
}
export default GifGridItem;