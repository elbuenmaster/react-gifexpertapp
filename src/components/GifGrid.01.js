import React, { useState, useEffect } from "react";
import { GifGridItem } from "../components/GifGridItem";
import {getGifs} from "../helpers/getGifs";

export const GifGrid = ({ category }) => {
  const apiKey = "eBHH7BLGeQviyhM8FyoIaVveWFeN72yl";
  const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(category)}&limit=10&api_key=${apiKey}`;

  const [images, setImages] = useState([]);

  

  useEffect(() => {
    getGifs(url)
    .then(imgs=>setImages(imgs));
  }, [url]);

  return (
    <>  
    <h3>{category}</h3>
    <div className='card-grid'>
    
      
        {images.map((imagen) => (
          <GifGridItem 
          key={imagen.id}
          {...imagen} />
        ))}
     
    </div>
   </> 
  );
};
