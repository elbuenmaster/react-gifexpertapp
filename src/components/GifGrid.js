import React, { useState, useEffect } from "react";
import { GifGridItem } from "../components/GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
 
  /* const [images, setImages] = useState([]);
  useEffect(() => {
    getGifs(url)
    .then(imgs=>setImages(imgs));
  }, [url]);
 */

  // al cargar la app por primera vez,data est√° vacio y loading esta en false
  const {data:images,loading}= useFetchGifs(category);
 
  return (
    <>  
    <h3>{category}</h3>
    {loading && <p>Cargando</p>}
   {   <div className='card-grid'>
    
      
        {images.map((imagen) => (
          <GifGridItem 
          key={imagen.id}
          {...imagen} />
        ))}
     
    </div> }
   </> 
  );
};

GifGrid.ProTypes={
  category:PropTypes.string.isRequired
}