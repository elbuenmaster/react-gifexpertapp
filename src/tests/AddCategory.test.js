import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { AddCategory } from "../components/AddCategory";
import PropTypes from "prop-types";

describe("AddCategory", () => {
  // creamos una funcion vacia para simular el parametro que recibe setCategorias
  // const setCategorias=()=>{}
  const setCategorias = jest.fn();
  //--  lo mismo, pero con jest.fn tenemos nuevas funcionalidades que ayudan al testing

  let wrapper;
  wrapper = shallow(<AddCategory setCategorias={setCategorias} />);
   
beforeEach(()=>{
     jest.clearAllMocks();
     wrapper = shallow(<AddCategory setCategorias={setCategorias} />);
   
})


  test("testing AddCategory", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("cambio de caja de texto", () => {
    const cajita = wrapper.find("input");
    const nuevovalor = "nuevo valor";
    cajita.simulate("change", {
      target: {
        value: nuevovalor,
      },
    });
    expect(wrapper.find("p").text().trim()).toBe(nuevovalor);
  });

  test("submit del formulario", () => {
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  });

  test("no debe postear si el valor de input está en blanco", () => {
    expect(setCategorias).not.toHaveBeenCalled();
  });

  test('debe de llamar el setCategorias y limpiar la caja de texto',()=>{
        // 1. simular el inputChange
        // 2. simular el submit
        // 3. setCategorias se debe haber llamado
        // 4. el valor del input debe estar en ''  
        let val='hola';
        wrapper.find('input').simulate('change',{target:{ value:val }} );
        wrapper.find('form').simulate('submit',{ preventDefault:()=>{} });

        expect(setCategorias).toHaveBeenCalled();

        wrapper.find('input').simulate('change',{target:{value:''}});

        //también es válido: wrapper.find('input').prop('value')
         expect( wrapper.find('input').props().value ).toBe('');
  })

});
