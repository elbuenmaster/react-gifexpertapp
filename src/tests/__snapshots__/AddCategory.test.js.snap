// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory testing AddCategory 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
